kernel/src/processbuffer.rs

```diff
@@ -148,7 +148,12 @@ unsafe fn raw_processbuf_to_rwprocessslice<'a>(
 /// This trait can be used to gain read-only access to memory regions
 /// wrapped in either a [`ReadOnlyProcessBuffer`] or a
 /// [`ReadWriteProcessBuffer`] type.
-pub trait ReadableProcessBuffer {
+///
+/// This is an `unsafe trait` as users of this trait need to trust that the
+/// implementation of [`ReadableProcessBuffer::ptr`] is correct, see [1].
```

Comment:
Okay, I read this and [the motivating issue](https://github.com/tock/tock/pull/4231) and the [linked documentation](https://doc.rust-lang.org/nomicon/safe-unsafe-meaning.html) _many_ times, and I think I finally understand the problem. However, I am slightly unsure of whether this as-is wholly _fixes_ the issue, or in part just _moves_ the problem.

The `ptr()` method in this trait returns a raw pointer. Currently, any code trying to do anything meaningful with that would need to use `unsafe` and verify the validity [and IIUC, the point of marking the trait `unsafe` here is to tell the _other_ `unsafe` code which uses the `ptr()` method that the ptr is valid, for some safety definition]. 
 - i.e., the current bug is that the `unsafe` implementation of `kernel::ipc::schedule_upcall` uses `slice.ptr()` without checking its validity

So what checks/enforces the validity of `ReadableProcessBuffer::ptr()` currently? As far as I can tell, the answer is `Read[Only|Write]ProcessBuffer::new[_external]`. That subtlety needs to be explicitly documented on lines 316 and 523, i.e., that the _reason_ the `unsafe impl` is safe is because the constructor for the type the `impl` is for provides for {invariants}.

Similarly, I think this needs to say what are the invariants for `ptr`, `len`, `enter` that an implementer of this `unsafe` trait must ensure hold?

---

Summarizing, I think #3757 observes that we have code which is marked unsafe that isn't properly (a) checking its invariants, or (b) documenting why it is safe to elide the checks. With just this change, we _still_ have the same bug, it's just _where_ the missing 'documentation for what it is safe to elide checks' has moved. For that reason, I don't think it's fair to say this "Fixes #3757".

I could be convinced that this is _better_ than what exists, so we should merge this and open a new issue more directly about the missing safety documentation. But, I'm not really convinced there's a huge amount of value in sealing this interface that isn't used incorrectly currently anyway without also explaining how to use it correctly.